libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +630,246 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+			(*it)->name() +
+			". Function with the same name and parameter types defined in two or more base classes."
+		);
+	}
+}
+
+set<ContractDefinition const*> ContractLevelChecker::resolveOverrideList(OverrideSpecifier const& _overrides) const
+{
+	set<ContractDefinition const*> resolved;
+
+	for (ASTPointer<UserDefinedTypeName> const& override: _overrides.overrides())
+	{
+		Declaration const* decl  = override->annotation().referencedDeclaration;
+		solAssert(decl, "Expected declaration to be resolved.");
+
+		// If it's not a contract it will be caught
+		// in the reference resolver
+		if (ContractDefinition const* contract = dynamic_cast<decltype(contract)>(decl))
+			resolved.insert(contract);
+	}
+
+	return resolved;
+}
+
+
+void ContractLevelChecker::checkModifierOverrides(FunctionSet const& _funcSet, ModifierSet const& _modSet, std::vector<ModifierDefinition const*> _modifiers)
+{
+	for (ModifierDefinition const* modifier: _modifiers)
+	{
+		if (contains_if(_funcSet, MatchByName{modifier->name()}))
+			m_errorReporter.typeError(
+				modifier->location(),
+				"Override changes function to modifier."
+			);
+
+		auto [begin,end] = _modSet.equal_range(modifier);
+
+		// Skip if no modifiers found in bases
+		if (begin == end)
+			continue;
+
+		if (!modifier->overrides())
+			overrideError(*modifier, **begin, "Overriding modifier is missing 'override' specifier.");
+
+		for (; begin != end; begin++)
+			if (ModifierType(**begin) != ModifierType(*modifier))
+				m_errorReporter.typeError(
+					modifier->location(),
+					"Override changes modifier signature."
+			);
+	}
+
+}
+
+void ContractLevelChecker::checkOverrideList(FunctionSet const& _funcSet, FunctionDefinition const& _function)
+{
+	set<ContractDefinition const*> specifiedContracts =
+		_function.overrides() ?
+		resolveOverrideList(*_function.overrides()) :
+		set<ContractDefinition const*>{};
+
+	// Check for duplicates in override list
+	if (_function.overrides() && specifiedContracts.size() != _function.overrides()->overrides().size())
+	{
+		// Sort by name to find duplicate for error reporting
+		vector<ASTPointer<UserDefinedTypeName>> list =
+			sortByName(_function.overrides()->overrides());
+
+		// Find duplicates and output error
+		for (size_t i = 1; i < list.size(); i++)
+			if (list[i]->namePath() == list[i-1]->namePath())
+			{
+				SecondarySourceLocation ssl;
+				ssl.append("First occurrence here: ", list[i-1]->location());
+				m_errorReporter.typeError(
+					list[i]->location(),
+					ssl,
+					"Duplicate contract " +
+						joinHumanReadable(list[i]->namePath(), ".") +
+						" found in override list of " +
+						_function.name() +
+						"."
+				);
+			}
+	}
+
+	decltype(specifiedContracts) missingContracts;
+
+	// Iterate over the overrides
+	for (auto [begin, end] = _funcSet.equal_range(&_function); begin != end; begin++)
+	{
+		// Validate the override
+		if (!checkFunctionOverride(_function, **begin))
+			break;
+
+		auto const result = find(
```

Comment:
Instead of these ten lines of code, it might be better to build up a set of contracts called 'expectedContracts' containing all `(*begin)->annotation().contract` and then after the loop, use `auto missingContracts = expectedContracts - specifiedContracts` and `invalidContracts = specifiedContracts - expectedContracts`.